runtime-transaction/src/instruction_details.rs

```diff
@@ -0,0 +1,273 @@
+use {
+    crate::{builtin_instruction_details::*, compute_budget_instruction_details::*},
+    solana_compute_budget::compute_budget_limits::*,
+    solana_sdk::{
+        instruction::CompiledInstruction,
+        pubkey::Pubkey,
+        transaction::{Result, TransactionError},
+    },
+    std::num::NonZeroU32,
+};
+
+/// Information about instructions gathered after scan over transaction;
+/// These are "raw" information that suitable for cache and reuse.
+#[cfg_attr(test, derive(Eq, PartialEq))]
+#[derive(Default, Debug)]
+pub struct InstructionDetails {
+    builtin_instruction_details: BuiltinInstructionDetails,
+    compute_budget_instruction_details: ComputeBudgetInstructionDetails,
+}
+
+impl InstructionDetails {
+    pub fn try_from<'a>(
+        instructions: impl Iterator<Item = (&'a Pubkey, &'a CompiledInstruction)>,
+    ) -> Result<Self> {
+        let mut compute_budget_instruction_details = ComputeBudgetInstructionDetails::default();
+        let mut builtin_instruction_details = BuiltinInstructionDetails::default();
+
+        for (i, (program_id, instruction)) in instructions.enumerate() {
+            compute_budget_instruction_details.process_instruction(
+                i as u8,
+                program_id,
+                instruction,
+            )?;
+            builtin_instruction_details.process_instruction(program_id, instruction)?;
+        }
+
+        Ok(Self {
+            builtin_instruction_details,
+            compute_budget_instruction_details,
+        })
+    }
+
+    pub fn sanitize_and_convert_to_compute_budget_limits(&self) -> Result<ComputeBudgetLimits> {
+        // Sanitize requested heap size
+        let updated_heap_bytes = self
+            .compute_budget_instruction_details
+            .requested_heap_size
+            .map_or(MIN_HEAP_FRAME_BYTES, |(_index, requested_heap_size)| {
+                requested_heap_size
+            })
+            .min(MAX_HEAP_FRAME_BYTES);
+
+        // Calculate compute unit limit
+        let compute_unit_limit = self
+            .compute_budget_instruction_details
+            .requested_compute_unit_limit
+            .map_or_else(
+                || {
+                    // NOTE: to match current behavior of:
+                    // num_non_compute_budget_instructions * DEFAULT
+                    self.builtin_instruction_details
+                        .count_builtin_instructions
+                        .saturating_add(
+                            self.builtin_instruction_details
+                                .count_non_builtin_instructions,
+                        )
+                        .saturating_sub(
+                            self.compute_budget_instruction_details
+                                .count_compute_budget_instructions,
+                        )
+                        .saturating_mul(DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT)
```

Comment:
nit: I'd prefer this to be pulled into a private helper function `fn num_non_compute_budget_instructions` on `InstructionDetails`

Reply:
0289d75e2ce84c9bb0d66b1591489c69f345935a does it